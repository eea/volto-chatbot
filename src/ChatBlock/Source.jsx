import { Popup } from 'semantic-ui-react';
import { SVGIcon } from './utils';

import FileIcon from './../icons/file.svg';
import GlobeIcon from './../icons/globe.svg';

import { injectLazyLibs } from '@plone/volto/helpers/Loadable/Loadable';

const SourceDetails_ = ({ source, index, luxon }) => {
  const {
    link,
    blurb,
    updated_at,
    source_type,
    semantic_identifier = 'untitled document',
  } = source || {};
  const parsedDate = updated_at ? luxon.DateTime.fromISO(updated_at) : null;
  const relativeTime = parsedDate?.toRelative();
  const isLinkType = source_type === 'web';
  const isDocumentType = source_type === 'file';

  const renderIcon = () => {
    if (isLinkType) {
      return <SVGIcon name={GlobeIcon} size="15" alt="Web icon" />;
    }
    if (isDocumentType) {
      return <SVGIcon name={FileIcon} size="15" alt="File icon" />;
    }
    return null;
  };

  const sourceContent = (
    <>
      {blurb && (
        <div className="source-desc">
          <span>{blurb}</span>
        </div>
      )}
      {updated_at && (
        <div className="source-date">
          <span>{relativeTime}</span>
        </div>
      )}
    </>
  );

  return (
    <>
      {isLinkType ? (
        <a
          href={link}
          rel="noreferrer"
          target="_blank"
          className="source source-link"
        >
          <div className="source-header">
            <span className="chat-citation">{index}</span>
            <div className="source-title" title={semantic_identifier}>
              {semantic_identifier}
            </div>
            {renderIcon()}
          </div>
          {sourceContent}
        </a>
      ) : (
        <div className="source">
          <div className="source-header">
            <Popup
              on="click"
              wide="very"
              content="This doc doesn't have a link."
              trigger={<span className="chat-citation">{index}</span>}
              popper={{ id: 'chat-citation-popup' }}
            />
            <div className="source-title" title={semantic_identifier}>
              {semantic_identifier}
            </div>
            {renderIcon()}
          </div>
          {sourceContent}
        </div>
      )}
    </>
  );
};

export const SourceDetails = injectLazyLibs(['luxon'])(SourceDetails_);
