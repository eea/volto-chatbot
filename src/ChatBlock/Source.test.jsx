import { MemoryRouter } from 'react-router-dom';
import configureStore from 'redux-mock-store';
import renderer from 'react-test-renderer';

import '@testing-library/jest-dom/extend-expect';
import { Provider } from 'react-intl-redux';
import { SourceDetails } from './Source';

const mockStore = configureStore();

jest.mock('@plone/volto/helpers/Loadable/Loadable', () => ({
  injectLazyLibs: () => (Component) => (props) => (
    <Component {...props} luxon={require('luxon')} />
  ),
}));

describe('SourceDetails', () => {
  it('should render the component with link type', () => {
    const store = mockStore({
      userSession: { token: '1234' },
      intl: {
        locale: 'en',
        messages: {},
      },
    });

    const props = {
      index: '1',
      source: {
        blurb: 'Vestibulum purus quam scelerisque ut',
        link: 'https://www.example.com',
        source_type: 'web',
        semantic_identifier: 'Nam at tortor in tellus',
        updated_at: null,
      },
    };

    const component = renderer.create(
      <Provider store={store}>
        <MemoryRouter>
          <SourceDetails {...props} />
        </MemoryRouter>
      </Provider>,
    );
    const json = component.toJSON();
    expect(json).toMatchSnapshot();
  });

  it('should render the component with doc type', () => {
    const store = mockStore({
      userSession: { token: '1234' },
      intl: {
        locale: 'en',
        messages: {},
      },
    });

    const props = {
      index: '2',
      source: {
        blurb: 'Vestibulum purus quam scelerisque ut',
        link: 'https://www.example.com',
        source_type: 'file',
        semantic_identifier: 'Nam at tortor in tellus',
        updated_at: null,
      },
    };

    const component = renderer.create(
      <Provider store={store}>
        <MemoryRouter>
          <SourceDetails {...props} />
        </MemoryRouter>
      </Provider>,
    );
    const json = component.toJSON();
    expect(json).toMatchSnapshot();
  });
});
