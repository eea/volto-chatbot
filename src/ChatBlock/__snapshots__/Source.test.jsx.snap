// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SourceDetails should render the component with doc type 1`] = `
<div
  className="source"
>
  <div
    className="source-header"
  >
    <span
      className="chat-citation"
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      2
    </span>
    <div
      className="source-title"
    >
      Nam at tortor in tellus
    </div>
    <svg
      className="icon"
      dangerouslySetInnerHTML={
        Object {
          "__html": undefined,
        }
      }
      fill="currentColor"
      height="18"
      stroke="currentColor"
      viewBox=""
      width="18"
      xmlns=""
    />
  </div>
  <div
    className="source-date"
  >
    <span>
      6 months ago
    </span>
  </div>
  <div
    className="source-desc"
  >
    <span>
      Vestibulum purus quam scelerisque ut
    </span>
  </div>
</div>
`;

exports[`SourceDetails should render the component with link type 1`] = `
<div
  className="source"
>
  <div
    className="source-header"
  >
    <a
      href="https://www.example.com"
      rel="noreferrer"
      target="_blank"
    >
      <span
        className="chat-citation"
      >
        1
      </span>
      <div
        className="source-title"
      >
        Nam at tortor in tellus
      </div>
      <svg
        className="icon"
        dangerouslySetInnerHTML={
          Object {
            "__html": undefined,
          }
        }
        fill="currentColor"
        height="18"
        stroke="currentColor"
        viewBox=""
        width="18"
        xmlns=""
      />
    </a>
  </div>
  <div
    className="source-desc"
  >
    <span>
      Vestibulum purus quam scelerisque ut
    </span>
  </div>
</div>
`;
